### PR TITLE
Bulk Editing: Add selection state to products list

### PR DESCRIPTION
**Pull request by ealeksandrov**
Part of: https://github.com/woocommerce/woocommerce-ios/issues/8517

## Description

This PR adds bulk editing state on Products List.
It replaces navbar content, disables pull-to-refresh, switches table view into multiple selection mode.

## Testing

1. Build and run the app in debug/alpha mode.
2. On the products list tap the "Bulk Edit" button in the navbar.
3. Confirm navbar title changes, all buttons are replaced with "Cancel".
4. Confirm pull-to-refresh is unavailable.
5. Select a few products from the list.
6. Confirm title changes, displaying amount of selected items.
7. Tap "Cancel" in navbar.
8. Confirm normal navbar title and items are back, pull-to-refresh is enabled.
 
## Video

https://user-images.githubusercontent.com/3132438/210849180-41e04008-4ee8-4e4e-bf89-47b2a9d1131f.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift [hunk 1]
```diff
@@ -345,6 +357,26 @@ private extension ProductsViewController {
         navigationItem.rightBarButtonItems = rightBarButtonItems
     }
 
+    func configureNavigationBarForEditing() {
+        configureNavigationBarTitleForEditing()
+        configureNavigationBarRightButtonItemsForEditing()
+    }
+
+    func configureNavigationBarTitleForEditing() {
+        let selectedProducts = tableView.indexPathsForSelectedRows?.count ?? 0
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift]: I'm not sure if we can use `tableView.indexPathsForSelectedRows` safely. When the table view reloads due to a page being loaded this property gets cleaned.


https://user-images.githubusercontent.com/562080/210893792-06fcf688-9cd1-42fa-87ad-4d90bf348b04.mov

I think we will have to change how we update the table view with new pages or we will have to store these selections on a view model.
 
What do you think?

Comment (ealeksandrov) [reply WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift]: Good catch! Added `ProductsListViewModel` to store selection state.

File: WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    func productIsSelected(_ productToCheck: Product) -> Bool {
+        return selectedProducts.contains(productToCheck)
+    }
+
+    func selectProduct(_ selectedProduct: Product) {
+        selectedProducts.insert(selectedProduct)
+//        if selectedProducts.contains(selectedProduct) {
+//        } else {
+//        }
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/ProductsListViewModel.swift]:
```suggestion
```

**Review by Ecarrion — APPROVED** [review]
Looks good! I just found one strange glitch where the cell separators disappear if you start selecting from the top of the list.

Feel free to create an issue for it!

Screenshot 2023-01-10 at 9 56 35 AM

**Comment by ealeksandrov** [comment]
Thanks! Created https://github.com/woocommerce/woocommerce-ios/issues/8608

**Pull request merged**